File: frontend/src/api/course.js
import request from './request'

// ==================== 课程 API ====================

/**
 * 课程列表（公开）
 * @param {Object} params - { page, page_size, keyword, category_id, status, teacher_id }
 */
export function getCourseList(params) {
  return request.get('/courses/', { params })
}

/**
 * 我的课程（讲师/管理员）
 */
export function getMyCourses(params) {
  return request.get('/courses/my', { params })
}

/**
 * 获取课程详情
 */
export function getCourseDetail(id) {
  return request.get(`/courses/${id}`)
}

/**
 * 创建课程
 */
export function createCourse(data) {
  return request.post('/courses/', data)
}

/**
 * 更新课程
 */
export function updateCourse(id, data) {
  return request.put(`/courses/${id}`, data)
}

/**
 * 删除课程
 */
export function deleteCourse(id) {
  return request.delete(`/courses/${id}`)
}

// ==================== 章节 API ====================

/**
 * 获取章节目录（树形）
 */
export function getChapterTree(courseId) {
  return request.get(`/courses/${courseId}/chapters`)
}

/**
 * 创建章节
 */
export function createChapter(courseId, data) {
  return request.post(`/courses/${courseId}/chapters`, data)
}

/**
 * 更新章节
 */
export function updateChapter(chapterId, data) {
  return request.put(`/chapters/${chapterId}`, data)
}

/**
 * 删除章节
 */
export function deleteChapter(chapterId) {
  return request.delete(`/chapters/${chapterId}`)
}

/**
 * 章节排序
 */
export function sortChapter(chapterId, sortOrder) {
  return request.put(`/chapters/${chapterId}/sort`, null, {
    params: { sort_order: sortOrder },
  })
}

File: backend/app/services/course.py
"""课程管理业务逻辑"""
from typing import Optional

from fastapi import HTTPException, status
from sqlalchemy.orm import Session, joinedload
from sqlalchemy import and_

from app.models.course import Course
from app.models.chapter import Chapter
from app.models.user import SysUser
from app.schemas.course import CourseCreate, CourseUpdate


class CourseService:

    @staticmethod
    def list_courses(
        db: Session,
        page: int = 1,
        page_size: int = 12,
        keyword: Optional[str] = None,
        category_id: Optional[int] = None,
        status: Optional[str] = "published",
        teacher_id: Optional[int] = None,
    ):
        """分页查询课程列表"""
        query = db.query(Course)

        # 公开列表只看已发布课程
        if status:
            query = query.filter(Course.status == status)

        if keyword:
            query = query.filter(Course.title.contains(keyword))
        if category_id:
            query = query.filter(Course.category_id == category_id)
        if teacher_id:
            query = query.filter(Course.teacher_id == teacher_id)

        total = query.count()
        items = (
            query.options(joinedload(Course.teacher))
            .order_by(Course.updated_at.desc())
            .offset((page - 1) * page_size)
            .limit(page_size)
            .all()
        )
        return {"total": total, "items": items}

    @staticmethod
    def get_course(db: Session, course_id: int) -> Course:
        course = (
            db.query(Course)
            .options(joinedload(Course.teacher), joinedload(Course.chapters))
            .filter(Course.id == course_id)
            .first()
        )
        if not course:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="课程不存在")
        return course

    @staticmethod
    def create_course(db: Session, data: CourseCreate, teacher_id: int) -> Course:
        course = Course(
            title=data.title,
            description=data.description,
            cover_image=data.cover_image,
            category_id=data.category_id,
            teacher_id=teacher_id,
            price=data.price,
            status=data.status,
        )
        db.add(course)
        db.commit()
        db.refresh(course)
        return course

    @staticmethod
    def update_course(db: Session, course_id: int, data: CourseUpdate, current_user: SysUser) -> Course:
        course = CourseService.get_course(db, course_id)

        # 权限：仅教师本人或管理员
        if current_user.role != "admin" and course.teacher_id != current_user.id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="无权修改此课程")

        update_data = data.model_dump(exclude_unset=True)
        for field, value in update_data.items():
            setattr(course, field, value)

        db.commit()
        db.refresh(course)
        return course

    @staticmethod
    def delete_course(db: Session, course_id: int, current_user: SysUser) -> None:
        course = CourseService.get_course(db, course_id)

        if current_user.role != "admin" and course.teacher_id != current_user.id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="无权删除此课程")

        db.delete(course)
        db.commit()
